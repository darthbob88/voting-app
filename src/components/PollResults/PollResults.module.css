form {
  border: 1px solid yellow;
  margin: 1rem;
}
