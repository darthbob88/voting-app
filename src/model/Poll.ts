export type Poll = {
    ID: string,
    name: string,
    creator: string,
    creation: number,
    expiration: number,
    method: string,
    candidates: string[],
    ballots: Ballot[]
}

export type Ballot = {
    ballotID: string,
    voterID: string,
    pollID: string,
    votes: any,
}