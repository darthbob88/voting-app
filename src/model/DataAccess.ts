import { Ballot, Poll } from "./Poll";

/*
 * Using local storage, just so I don't have to set up
 * Firebase until I'm more confident in the data structures. 
 */
//TODO: Set up Firebase, or whatever other data storage I decide to use.

export const createPoll = (newPoll: Poll) => {
    if (localStorage.getItem(newPoll.ID) != null) {
        throw new Error(`Invalid poll ${newPoll.ID}`);
    }
    localStorage.setItem(newPoll.ID, JSON.stringify(newPoll));
    return newPoll.ID;
}

export const retrievePoll = (pollID: string): Poll | null => {
    const jsonBlob = localStorage.getItem(pollID);
    if (jsonBlob == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }

    return JSON.parse(jsonBlob);

}

export const retrievePollList = (): string[] => {
    return Object.keys(localStorage);
}

export const castBallot = (newBallot: Ballot, pollID: string) => {
    const poll = retrievePoll(pollID);
    if (poll == null) {
        throw new Error(`Invalid poll ${pollID}`);
    }
    poll.ballots.push(newBallot);
    localStorage.setItem(poll.ID, JSON.stringify(poll));
}