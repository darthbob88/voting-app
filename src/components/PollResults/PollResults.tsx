import { retrievePoll } from '../../model/DataAccess';
import { Ballot } from '../../model/Poll';

import styles from "./PollResults.module.css";

type BallotProps = { pollID: string };

function PollResults({ pollID }: BallotProps) {

    const poll = retrievePoll(pollID);
    if (null == poll) {
        return <div>Invalid poll ID {pollID}</div>;
    }


    const renderResults = (ballots: Ballot[], method: string, options: string[]) => {
        if (method === "FPTP") {
            const result = new Map<string, number>();
            options.forEach(option => result.set(option, 0));
            let curLeader = options[0];
            ballots.reduce((acc, cur) => {
                const choice = cur.votes as string;
                if (!options.includes( choice))
                    throw new Error(`Invalid ballot: ${choice} is not a valid option`);
                    
                const currVal = acc.get(choice) ?? 0;
                acc.set(choice, currVal + 1);
                if (acc.has(curLeader) && currVal > (acc.get(curLeader) ?? 0)) {
                    curLeader = choice;
                }
                return acc;
            }, result);
            return <div className={styles.ballot}>
                Winner is {curLeader} with {result.get(curLeader)} votes out of {ballots.length}
                <br />
                Detailed result: 
                <ol>
                {
                    [...result.entries()].sort((a, b) => b[1] - a[1]).map(candidate => <li key={candidate[0]}>{candidate[0]} - {candidate[1]} votes</li>)
                }
                </ol>
            </div>
        } else if (method === "IRV") {
            return <div className={styles.ballot}> A series of dropdown menus to select the first/second/third/etc candidates</div>
        } else {
            return null;
        }
    }



    return (
        <div>
            <h3>Results for {poll.name}</h3>
            {renderResults(poll.ballots, poll.method, poll.candidates)}
        </div>
    );
}

export default PollResults;