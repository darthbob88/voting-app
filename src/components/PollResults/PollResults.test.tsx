import { render } from '@testing-library/react';
import { createPoll } from '../../model/DataAccess';
import { testPoll } from '../../model/Poll';
import PollResults from './PollResults';

// TODO: Again, clear this once I have a proper database.
beforeEach(() => {
    localStorage.clear();
    createPoll(testPoll);

})

test('renders PollResults component', () => {
    const { getByText, container } = render(<PollResults pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});

xtest('user can cast ballot with proper result', () => {
    const { getByText, container } = render(<PollResults pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});

xtest('user cannot submit empty ballot', () => {
    const { getByText, container } = render(<PollResults pollID={testPoll.ID} />);
    expect(container).toMatchSnapshot();
});



