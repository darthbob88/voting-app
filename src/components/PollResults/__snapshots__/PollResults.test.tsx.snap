// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PollResults component 1`] = `
<div>
  <div>
    <h3>
      Results for 
      pollName
    </h3>
    <div
      class="ballot"
    >
      Winner is 
      apple
       with 
      0
       votes out of 
      0
      <br />
      Detailed result:
      <ol>
        <li>
          apple
           - 
          0
           votes
        </li>
        <li>
          orange
           - 
          0
           votes
        </li>
        <li>
          banana
           - 
          0
           votes
        </li>
      </ol>
    </div>
  </div>
</div>
`;
